src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs

```diff
@@ -244,7 +244,7 @@ private OpenApiSchema GenerateConcreteSchema(DataContract dataContract, SchemaRe
                 case DataType.Object:
                     {
                         schemaFactory = () => CreateObjectSchema(dataContract, schemaRepository);
-                        returnAsReference = true;
+                        returnAsReference = !_generatorOptions.UseInlineDefinitionsForObjects;
```

Comment:
From the original PR for this change #2426, you'll want to add support for self-referencing types:

```csharp
returnAsReference = !_generatorOptions.UseInlineDefinitionsForObjects ||
                                        dataContract.ObjectProperties.Any(prop =>
                                            prop.MemberType.IsAssignableFrom(dataContract.UnderlyingType) ||
                                            prop.MemberType.GenericTypeArguments.Any(type =>
                                                type.IsAssignableFrom(dataContract.UnderlyingType)));
```

Reply:
The test for this to also validate the change:

```csharp
        [Fact]
        public void GenerateSchema_SupportsOption_UseReferenceForSelfReferenceTypesWhenUseInlineDefinitionsForObjects()
        {
            var subject = Subject(
                configureGenerator: c => c.UseInlineDefinitionsForObjects = true
            );

            var schema = subject.GenerateSchema(typeof(SelfReferencingType), new SchemaRepository());

            Assert.NotNull(schema.Reference);
            Assert.Equal("SelfReferencingType", schema.Reference.Id);
        }
```